2) YALB-84: Primary-nav

ModulesUnraveled opened the pull request:
## [YALB-84: Primary-nav](https://yaleits.atlassian.net/browse/YALB-84)

### NOTES:
- DONE ~The toggle arrows will be replaced with icons once that PR is done. For now, it's a simple caret (so the alignment is off)~
- There are some other known design tweaks Ashley and I will make next week (font sizes, etc) I don't think any of them should prevent other PR reviews from happening though.

### Description of work
- Creates the Primary Nav component

### Testing Link(s)
- [ ] Navigate to the [Primary Nav Story](https://deploy-preview-46--dev-component-library-twig.netlify.app/?path=/story/organisms-menu-primary-nav--primary-nav) to see it in isolation
- [ ] Navigate to the [Any of the example page stories](https://deploy-preview-46--dev-component-library-twig.netlify.app/?path=/story/page-examples-page-examples--standard-page) to see it in context of the site header/page

### Functional Review Steps
- [ ] Verify everything looks and functions as expected above and below 992px

### Design Review
- [ ] Verify the designs match the [Figma Designs](https://www.figma.com/file/OVfmCW5s1gI7xnrM8ibX2y/UI-Kit-Alpha-%5BYaleSites%5D)

### Accessibility Review
- [ ] Verify the component meets Accessibility requirements

kara-franco reviewed (changes requested):
Mobile menu keyboard accessibility is perfect! Also, the menu levels are announced really nicely in VoiceOver and NVDA screen readers. 

Only issue I caught is that the drop down menu remains open when tabbing to the next Level 1 (ex. menu stays open when I tab to Academics). The expected behavior is to close the drop down when the user tabs out of it.  

<img width="840" alt="Screen Shot 2022-03-22 at 6 53 44 PM" src="https://user-images.githubusercontent.com/12991593/159589644-be593c01-fafb-488c-972f-4c7ee2e3d447.png">

kara-franco commented:
re: @ahitson design change to remove focus indicators on mouse click- there is a nifty polyfill available called focus-visible that can solve this issue relatively easily. We used it in our component library: 

- [https://yale-a11y.gitlab.io/ui-component-library/documentation#version-history-and-changelog#focus-visible](https://yale-a11y.gitlab.io/ui-component-library/documentation#version-history-and-changelog#focus-visible)
- [https://github.com/WICG/focus-visible](https://github.com/WICG/focus-visible)

You may already know this, but wanted to share!

ModulesUnraveled commented:
@kara-franco Thanks for pointing me in that direction, I've used `:focus` and `:focus-within`, but hadn't really explored `:focus-visible` yet. [It looks like it's supported](https://caniuse.com/?search=focus-visible) (as a selector) in every browser that we support except Safari at this point.

So, I'd recommend that instead of implementing a polyfill, we just use the selector according to the spec, and provide the small fallback mentioned in the polyfill's docs so that for browsers that don't support `:focus-visible` (namely Safari) use browser styles, but everything else will use our "designed" styles.

I just checked on Chrome and Safari 15.3 locally, and both browsers seem to work as expected (Chrome uses our styles, and Safari uses browser styles.)

I'm making more changes now, but I just pushed that up. So, if you want to take a look at that, and let me know if it seems to work in all of your environments, I think that's the way we should go, personally.

kara-franco commented:
> @kara-franco Thanks for pointing me in that direction, I've used `:focus` and `:focus-within`, but hadn't really explored `:focus-visible` yet. [It looks like it's supported](https://caniuse.com/?search=focus-visible) (as a selector) in every browser that we support except Safari at this point.
> 
> So, I'd recommend that instead of implementing a polyfill, we just use the selector according to the spec, and provide the small fallback mentioned in the polyfill's docs so that for browsers that don't support `:focus-visible` (namely Safari) use browser styles, but everything else will use our "designed" styles.
> 
> I just checked on Chrome and Safari 15.3 locally, and both browsers seem to work as expected (Chrome uses our styles, and Safari uses browser styles.)
> 
> I'm making more changes now, but I just pushed that up. So, if you want to take a look at that, and let me know if it seems to work in all of your environments, I think that's the way we should go, personally.

Awesome!!! So I just updated to the latest version of Safari and it now supports focus-visible, so I think the above solution is perfect. 

I did notice something odd in Firefox with the `<button>` focus-visible styling, it's being overridden with a 1px dotted gray indicator. 

<img width="386" alt="Screen Shot 2022-03-23 at 1 47 11 PM" src="https://user-images.githubusercontent.com/12991593/159764262-2cb723b3-146f-4922-8efd-3edfdcea691f.png">

ModulesUnraveled commented:
@kara-franco It looks like that gray dotted outline is a FireFox-specific reset that ships with normalize... I tried to get rid of it, but then it broke the styles in other browsers... is it something we can live with? If not, I think we should create a bug-fix ticket to address it later. I don't think it's something I should spin my wheels on at this time.

I think your other issues are addressed.

kara-franco commented:
> @kara-franco It looks like that gray dotted outline is a FireFox-specific reset that ships with normalize... I tried to get rid of it, but then it broke the styles in other browsers... is it something we can live with? If not, I think we should create a bug-fix ticket to address it later. I don't think it's something I should spin my wheels on at this time.
> 
> I think your other issues are addressed.

@ModulesUnraveled Excellent, I've confirmed on Safari, Firefox and Chrome that the menu closes when tabbed out. 

I completely agree, I will log a bug-fix for the Firefox buttons.

ModulesUnraveled commented:
@ahitson I converted your comment to a check list, and addressed all of the items except one:
- As we talked about on Zoom, we'll wait for the new icons before we start fiddling with their sizes.

Ready for your re-review

ahitson commented:
So close! Please make the space between the parent link (2nd level) and first child (3rd level) equal to the space between the child pages.

On mobile, please bring up the first child in the dropdown closer to the parent.

Can we add a smidge of space between the main level menu items or does that put us in dangerous territory?

And bring down the menu so it's hitting the baseline of the logo (example below)
![Frame 456 (1)](https://user-images.githubusercontent.com/80782149/159970426-01778659-5099-4b3a-a0ab-7a043a2d69c5.png)

callinmullaney commented on components/02-molecules/menu/menu-toggle/menu-toggle.js:
```diff
@@ -1,42 +1,135 @@
 Drupal.behaviors.menuToggle = {
   attach(context) {
-    // Selectors
+    // Selectors.
```
Per our slack discussion you can likely trim down this file. 

- JS should calculate the the header height but assign that value to a CSS custom property.
- SCSS should use that custom property to determine the height of the mobile menu `height: calc(100vw - var(--heade-height);`
- Remove the media query JS. The frozen class should have no affect on desktop sizes which can be applied via a standard CSS media query

ModulesUnraveled replied:
I did all of this. 👍

callinmullaney commented on components/03-organisms/menu/primary-nav/_primary-nav.scss:
```diff
@@ -0,0 +1,201 @@
+  }
+}
+
+@mixin primary-nav-item-level-1 {
```
Not a blocker for merging but these mixins are only used once and are not very big.

ModulesUnraveled replied:
Done

callinmullaney reviewed (changes requested):
@ModulesUnraveled back to you

ModulesUnraveled commented:
@ahitson @callinmullaney Back to you two for reviews!

callinmullaney commented on components/03-organisms/site-header/_site-header.scss:
```diff
@@ -66,19 +82,19 @@ $site-header-animation-speed: 800ms;
     left: 0;
     background: var(--color-background);
-    max-height: 100vh;
+    max-height: calc(100vh - var(--header-height-mobile));
```
This is much more noticeable when the border thickness value is set to `8` but I believe this line should be `max-height: calc(100vh - var(--header-height-mobile) + var(--site-header-border-thickness));`

ModulesUnraveled replied:
I updated the css to keep the header border in sight. Hoping that's what you were trying to achieve.

callinmullaney reviewed (approved):
@ModulesUnraveled The JS file looks much better and is doing a lot less compared to before. Nice work! 

There is one last minor comment I left for you to address. Otherwise, this seems good to merge from my perspective.

ModulesUnraveled commented:
@ahitson I made the sub menus "slide" in and out on mobile.

@kara-franco Would you mind double checking the mobile nav now? I had to change some stuff to make the sliding work, but I used a methodology we've used elsewhere, so I think it should be accessible

kara-franco commented:
Very slick!! I tested in Chrome, Safari and Firefox, looks great, keyboard accessibility is very good. Sounds great in VO.
